tools/collect_hana_sr_diag.sh

```diff
@@ -0,0 +1,645 @@
+    fi
+}
+
+collect_traces() {
+    local node_ip=$1
+    local output_dir=$2
+    local trace_pattern=$3
+    local output_file="$output_dir/${trace_pattern}_trace.log"
+
+    log_message "INFO" "Collecting ${trace_pattern} traces"
+    local remote_command="sudo find /usr/sap/${SAP_SID}/HDB${INSTANCE_NUM}/ -name '${trace_pattern}_*.trc' -print0 | while IFS= read -r -d '' file; do echo -e \"\n\n=== \${file} ===\n\"; sudo tail -n 200 \"\${file}\"; done"
+    if check_and_run_remote_command "${node_ip}" "find" "${remote_command}" "${output_file}"; then
+        check_file "${output_file}"
+    fi
+}
+
+collect_listening_ports() {
+    local node_ip=$1
+    local output_dir=$2
+    local tool_name="ss"
+    local remote_command="sudo ss -tlnp"
+    local output_file="${output_dir}/ss_tlnp.txt"
+
+    log_message "INFO" "Collecting: Listening Ports (ss -tlnp)"
+    if check_and_run_remote_command "${node_ip}" "${tool_name}" "${remote_command}" "${output_file}" "${LOG_FILE}"; then
+        check_file "${output_file}"
+        return $? # Return the status of check_file
+    else
+        log_message "ERROR" "Failed to collect listening ports on ${node_ip}."
+        return 1
+    fi
+}
+
+detect_hana_replication_port() {
+    local node_ip=$1
+    local output_dir=$2
+    local ss_output_file="${output_dir}/ss_tlnp.txt"
+    local detected_port=""
+
+    if [[ -f "$ss_output_file" ]]; then
+        # Try to find an external IP for hdbnameserver first, exclude 127.0.0.1 and 0.0.0.0
+        detected_port=$(grep 'hdbnameserver' "$ss_output_file" | grep -v '127.0.0.1' | grep -v '0.0.0.0' | grep -oP ':\K\d+' | head -n 1)
+        # If not found, try 0.0.0.0
+        if [[ -z "$detected_port" ]]; then
+            detected_port=$(grep 'hdbnameserver' "$ss_output_file" | grep '0.0.0.0' | grep -oP ':\K\d+' | head -n 1)
+        fi
+        # If still not found, try localhost
+        if [[ -z "$detected_port" ]]; then
+            detected_port=$(grep 'hdbnameserver' "$ss_output_file" | grep '127.0.0.1' | grep -oP ':\K\d+' | head -n 1)
+        fi
+        # If still not found, log an error, but don't echo anything to stdout for the variable assignment
+        if [[ -z "$detected_port" ]]; then
+            log_message "ERROR" "Not possible to detect HANA replication port for $node_ip from $ss_output_file."
+        fi
+    else
+        log_message "ERROR" "Missing $ss_output_file for $node_ip. Cannot detect HANA replication port."
+    fi
+
+    echo "$detected_port" # Will be empty if port not found or file missing, preventing error messages from being assigned.
+}
+
+parse_inventory_file() {
+    local inventory_file="$1"
+    local primary_ip=""
+    local secondary_ip=""
+
+    if [[ -f "$inventory_file" ]]; then
+        primary_ip=$(grep -A 1 'vmhana01:' "$inventory_file" | grep 'ansible_host:' | awk '{print $2}' | tr -d "'\"")
+        secondary_ip=$(grep -A 1 'vmhana02:' "$inventory_file" | grep 'ansible_host:' | awk '{print $2}' | tr -d "'\"")
```

Comment:
```suggestion
        primary_ip=$(grep -A 1 'vmhana01:' "$inventory_file" | grep 'ansible_host:' | awk -F\" '{print $2}')
        secondary_ip=$(grep -A 1 'vmhana02:' "$inventory_file" | grep 'ansible_host:' | awk -F\" '{print $2}')
```
Also, pretty sure there's a way to combine both `grep`s and the `awk` in a single `awk`, but couldn't find how quickly.